DESIGN.md

```diff
@@ -0,0 +1,74 @@
+Sandboxing: [Step towards modular architecture in OpenSearch](https://github.com/opensearch-project/OpenSearch/issues/1422)  
+Security: [Security for extensions](SECURITY.md) 
+
+## Plugins Architecture
+
+![](Docs/plugins.png)
+
+Plugins are installed via [`opensearch-plugin`](https://github.com/opensearch-project/OpenSearch/blob/main/distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java) and are class loaded into OpenSearch.
+Plugins run within OpenSearch as a single process. Plugins interface with OpenSearch via extension points which plug into the core modules of OpenSearch.
+This [blog post](https://opensearch.org/blog/technical-post/2021/12/plugins-intro/) helps untangle how plugins work. 
+
+Walking through an example, a Plugin would like to register a custom setting which could be toggled via Rest API by the user. 
+The plugin uses compiles with OpenSearch x.y.z version and generates a `.zip`.   
+This `.zip` file is installed via `opensearch-plugin` tool which unpacks the code and places it under `~/plugins/<plugin-name>`.
+During the bootstrap of OpenSearch node, it class loads all the code under `~/plugins/` directory. `Node.java` makes a call to get all settings the plugins would like to register. These settings are used as `additionalSettings` and construct `SettingsModule` instance which tracks all settings.
+
+## Extensions Architecture
+
+![](Docs/Extensions.png)
+
+Extensions are independent processes which are built using `opensearch-sdk`. They communicate with OpenSearch via [transport](https://github.com/opensearch-project/OpenSearch/tree/main/modules/transport-netty4) protocol which today is used to communicate between OpenSearch nodes.  
+Extensions are designed to extend features via APIs which are exposed using extension points of OpenSearch.  
+
+### Discovery
+Extensions are discovered and configured via `extensions.yml` which is read by OpenSearch during the node bootstrap. `ExtensionsOrchestrator` reads through the config file at `~/extensions` and registers extensions within OpenSearch.
+Here is an example extension configuration `extensions.yml`:
+
+```
+extensions:
+  - name: opensearch-sdk
+    uniqueId: opensearch-sdk-1
+    hostName: 'sdk_host'
+    hostAddress: '127.0.0.1'
+    port: '4532'
+    version: '1.0'
+    description: Extension for the Opensearch SDK Repo
+    opensearchVersion: '3.0.0'
+    javaVersion: '14'
+    className: ExtensionsRunner
+    customFolderName: opensearch-sdk
+    hasNativeController: false	
+```
+
+
+### Communication
+As we are running extensions on the port defined in the `extensions.yml`, the communication between OpenSearch and Extensions happens using a ServerSocket which binds the port and the host address. OpenSearch will initialize the extensions during the bootstrap by making a request to all the extensions running on different ports and thus creating a medium for the future requests.
+
+### OpenSearch SDK
```

Comment:
This sections seems to be more about how to migrate plugins to extensions with this SDK, seems like it should be in another section or reshape to explain why there is an extension sdk.

Also I don't think this name `OpenSearch SDK` is useful - can we pick a better one like `opensearch-extension-sdk`? https://github.com/opensearch-project/opensearch-sdk/issues/13

Reply:
Also, nix the word `currently` the document is in the present tense.

Reply:
@owaiskazi19 could you take care of it?

Reply:
Should I update the section name to `Why OpenSearch SDK?`.
We can update the DESIGN.md with the right name of SDK repo once #13 is resolved.

Reply:
We could leave it for now. Since the issue is still not done. 
Whenever the renaming happens, lets make sure this doc is updated.